### PR TITLE
Adds spell checking

### PR DESCRIPTION
**Pull request by codingjoe**
What this PR adds: 
- without installed aspell / enchant the tests go as usual, all the translation spellchecking tests are just skipped. (you see the skipping then)
- when you install the corresponding aspell dictionaries, and enchant (see the new readme), the items are not skipped any more, and you should see the spelling errors
- you can set an environment variable to a directory, to provide private wordlists for each language (also see the readme)

### CODE REVIEW

File: pytest_translations.py [hunk 1]
```diff
@@ -147,6 +155,45 @@ def reportinfo(self):
         return (self.fspath, -1, "po-fuzzy")
 
 
+lang_dicts = [name for name, dict in enchant.list_dicts()]
```
Comment (syphar) [inline pytest_translations.py]: you are shadowing a python builtin (`dict`)

File: pytest_translations.py [hunk 2]
```diff
@@ -147,6 +155,45 @@ def reportinfo(self):
...
+class PoSpellCheckingItem(Item):
+
+    def __init__(self, line, lang_dict, name, parent):
+        self.line = line
+        self.lang_dict = lang_dict
+        super(PoSpellCheckingItem, self).__init__(name, parent)
+        self.add_marker(MARKER_NAME)
+
+    def runtest(self):
+        tokenizer = get_tokenizer(chunkers=[HTMLChunker],filters=(EmailFilter, URLFilter, WikiWordFilter))
```
Comment (syphar) [inline pytest_translations.py]: pep8 (space after comma)

**Comment by syphar** [comment]
- tests fail 
- IMHO the spellchecking should either be optional, or skip (with a warning) checking a language if the dictionary is not found.

File: pytest_translations.py [hunk 3]
```diff
@@ -1,17 +1,25 @@
...
 
-__version__ = '0.2.0'
+from enchant.tokenize import get_tokenizer, WikiWordFilter, URLFilter, EmailFilter, HTMLChunker
+
+__version__ = '1.0.0'
 
 MARKER_NAME = 'translations'
 
 
+PWL = os.environ.get('PYTEST_TRANSLATIONS_PRIVATE_WORD_LIST', None)
```
Comment (syphar) [inline pytest_translations.py]: shouldn't this be per language?

File: pytest_translations.py [hunk 4]
```diff
@@ -169,6 +216,19 @@ def collect(self):
             self.name,
             self,
         )
+        parsed = _open_po_file(self.fspath)
+        lang_dict = parsed.metadata.get('Language', 'en')
+        if lang_dict not in lang_dicts:
+            pytest.fail('Dictionary "%s" not found.' % lang_dict)
```
Comment (syphar) [inline pytest_translations.py]: `pytest.skip` ? or `pytest.xfail`?

File: README.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    - python2-enchant
+    - python3-enchant
+    - aspell-en
+    - aspell-de
+```
+
+On Mac you can use brew to install:
+
+```bash
+brew install pyenchant aspell-aspell-{en|de|CHOSE YOUR LANGUAGE CODES}
```
Comment (syphar) [inline README.md]: - `pyenchant` does not exist
- same as the `aspell` packages here

**Comment by syphar** [comment]
- tests fail
- couldn't test further, since I couldn't install the required packages.

**Comment by codingjoe** [comment]
Ok, I should have caught all your comments.

File: pytest_translations.py (hunk 3 above)
Comment (syphar) [inline pytest_translations.py]: ?

**Comment by codingjoe** [comment]
The dicts for multiple languages are still missing.

**Comment by codingjoe** [comment]
@syphar I added multiple language support for private word lists and python format string chunkers. Ready when you are :)

File: pytest_translations.py [hunk 6]
```diff
@@ -147,6 +158,141 @@ def reportinfo(self):
...
+        raise StopIteration()
+
+    def _is_tag(self,text,offset):
+        if offset+1 < len(text):
+            if text[offset+1].isalpha():
+                return True
+        return False
+
+
+class OldStringFormatChunker(Chunker):
```
Comment (syphar) [inline pytest_translations.py]: tests missing

and it's nearly a 100% duplication of the `StringFormatChunker`

File: pytest_translations.py [hunk 7]
```diff
@@ -147,6 +158,141 @@ def reportinfo(self):
...
+if enchant:
+    lang_dicts = [name for name, _ in enchant.list_dicts()]
+else:
+    lang_dicts = []
+
+
+class StringFormatChunker(Chunker):
+    """Chunker for breaking up HTML documents into chunks of checkable text.
+
+    The operation of this chunker is very simple - anything between a "<"
```
Comment (syphar) [inline pytest_translations.py]: copy/past docstring

File: test_translations.py [hunk 8]
```diff
@@ -109,7 +122,7 @@ def test_uses_argument(self, testdir):
             """
         )
         result = testdir.runpytest()
-        assert result.ret == 0
+        # assert result.ret == 0, result.stdout
```
Comment (syphar) [inline test_translations.py]: remove or fix

**Comment by syphar** [comment]
I ran spell-checking on a repo, and only got this

image

**Comment by syphar** [comment]
I also see 

```
SKIP [5] ~/src/pytest-translations/pytest_translations.py:322: Dictionary "" not found.
```

Reason was that the `Language` attribute in some of the `.po` files was missing. 

I think we should fail in that case. Or assume english. But I tend to fail.

File: pytest_translations.py [hunk 9]
```diff
@@ -169,6 +315,23 @@ def collect(self):
             self.name,
             self,
         )
+        if enchant:
+            parsed = _open_po_file(self.fspath)
+            lang_dict = parsed.metadata.get('Language', 'en')
+            if lang_dict not in lang_dicts:
+                pytest.skip('Dictionary "%s" not found.' % lang_dict)
+            if PWL:
```
Comment (syphar) [inline pytest_translations.py]: shouln't we check if the language-specific `PWL` is actually there?

File: pytest_translations.py [hunk 10]
```diff
@@ -169,6 +315,23 @@ def collect(self):
             self.name,
             self,
         )
+        if enchant:
+            parsed = _open_po_file(self.fspath)
+            lang_dict = parsed.metadata.get('Language', 'en')
```
Comment (syphar) [inline pytest_translations.py]: the naming could  be better. 

here `lang_dict` is the language string, 

later on it's replaced with the `DictWithPWL`

**Comment by syphar** [comment]
@codingjoe you can have a first look

**Comment by normansander** [comment]
A description would be helpful

**Comment by syphar** [comment]
@headione you mean about what the PR should change / do? 
or general documentation?

**Comment by normansander** [comment]
yes @syphar, i was missing a little description in this pull request and its impacts. for me it was a bit hard to understand the feature without it, although it has a precise title. but other than that LGTM 👍🏼

**Comment by syphar** [comment]
@headione added a description

File: .travis.yml [hunk 11]
```diff
@@ -1,16 +1,24 @@
...
+    - aspell-en
+    - aspell-de
+install:
+  - pip install -U pip
+  - pip install -r requirements-dev.txt
+  - pip install -U coveralls
+script:
+  - coverage run --source="pytest_translations.py" -m "pytest"
+after_success:
+  - coveralls
```
Comment (codingjoe) [inline .travis.yml]: @syphar new line missing

**Comment by codingjoe** [comment]
@syphar I'm happy except the missing new line? Is there anything you'd like to add before I merge?

**Comment by syphar** [comment]
@codingjoe newline fixed. 

If you tested and everything is fine, feel free to merge

**Pull request merged**